pkg/topology/full/full.go

```diff
@@ -20,81 +21,111 @@ func init() {
 	rand.Seed(time.Now().UnixNano())
 }
 
-var _ topology.Driver = (*driver)(nil)
+var _ topology.Driver = (*Driver)(nil)
 
-// driver drives the connectivity between nodes. It is a basic implementation of a connectivity driver.
+// Driver drives the connectivity between nodes. It is a basic implementation of a connectivity Driver.
 // that enabled full connectivity in the sense that:
-// - Every peer which is added to the driver gets broadcasted to every other peer regardless of its address.
+// - Every peer which is added to the Driver gets broadcasted to every other peer regardless of its address.
 // - A random peer is picked when asking for a peer to retrieve an arbitrary chunk (Peerer interface).
-type driver struct {
-	mtx         sync.Mutex
-	connected   map[string]swarm.Address
-	discovery   discovery.Driver
-	addressBook addressbook.Getter
+type Driver struct {
+	discovery     discovery.Driver
+	addressBook   addressbook.GetPutter
+	connecter     p2p.Connecter
+	receivedPeers map[string]bool
+	logger        logging.Logger
 }
 
-func New(disc discovery.Driver, addressBook addressbook.Getter) topology.Driver {
-	return &driver{
-		connected:   make(map[string]swarm.Address),
-		discovery:   disc,
-		addressBook: addressBook,
+func New(disc discovery.Driver, addressBook addressbook.GetPutter, connecter p2p.Connecter, logger logging.Logger) *Driver {
+	return &Driver{
+		discovery:     disc,
+		addressBook:   addressBook,
+		connecter:     connecter,
+		receivedPeers: make(map[string]bool),
+		logger:        logger,
 	}
 }
 
 // AddPeer adds a new peer to the topology driver.
 // The peer would be subsequently broadcasted to all connected peers.
 // All conneceted peers are also broadcasted to the new peer.
-func (d *driver) AddPeer(overlay swarm.Address) error {
-	d.mtx.Lock()
-	defer d.mtx.Unlock()
+func (d *Driver) AddPeer(addr swarm.Address) error {
+	if d.receivedPeers[addr.ByteString()] {
+		return nil
+	}
 
-	ma, exists := d.addressBook.Get(overlay)
+	d.receivedPeers[addr.ByteString()] = true
+	connectedPeers := d.connecter.Peers()
+	ma, exists := d.addressBook.Get(addr)
 	if !exists {
 		return topology.ErrNotFound
 	}
 
-	var connectedNodes []discovery.BroadcastRecord
-	for _, addressee := range d.connected {
-		cma, exists := d.addressBook.Get(addressee)
-		if !exists {
-			return topology.ErrNotFound
-		}
-
-		err := d.discovery.BroadcastPeers(context.Background(), addressee, discovery.BroadcastRecord{Overlay: overlay, Addr: ma})
+	if !isConnected(addr, connectedPeers) {
```

Comment:
`isConnected` is used only once. maybe inline the logic here?

Reply:
I think that isConnected is a nice function to have and makes the code in this function easier to read.